### ISSUE TITLE
DELETE request with body or 'Transfer-Encoding: chunked' header causes server to close connection 

### ISSUE THREAD

**Issue by PharkMillups** [webmachine/webmachine#82]
Moving this over from Bugzilla. The original ticket contents are [here](https://issues.basho.com/show_bug.cgi?id=1119). Please use this for all future discussion.

**Comment by b** [comment]
Code and instructions for minimal repro here https://github.com/b/test_disco

**Comment by b** [comment]
This situation occurs when Transfer-encoding: chunked and there is no body or an invalid body.  I don't see how such a request is valid.

**Comment by nathanaschbacher** [comment]
Thanks Ben for helping track this down on the WebMachine/MochiWeb side.  This seems to be regarding some inconsistency in the HTTP spec about how to deal with chunked/body on DELETE verbs.

You mentioned that this doesn't seem reasonably fixable in MochiWeb, so at least in node.js it's necessary to thwart the http module's default behavior by manually ensuring chunked encoding isn't used and that there's no content body.

The easiest way to do this is set a "Content-Length: 0" header on any outgoing DELETE request and making sure you're not passing anything into req.write() or req.end() for the body.

That solved the issue for our Riak client.  Now we can run gobs of concurrent DELETE requests just like we could HEAD, GET, PUT, POST requests.

I'll make a PR to the node.js repo that fixes this directly, but I don't know if it will be adopted.

**Comment by slfritchie** [comment]
On the "wrd-content-length-with-streaming" branch, which I hope is not too divergent from the "master" branch (no, I didn't look)....

It appears that webmachine_decision_core.erl is hardcoded so that `delete_resource/2` must return `true`, otherwise we'll give a 500 status to the client:

```
decision(v3m20) ->
    decision_test(resource_call(delete_resource), true, v3m20b, 500);
```

Also, if `delete_resource/2` returns `{halt, integer()}`, then I see an exception like this one (sorry about the bad formatting):

```
[{webmachine_resource,do,3,[{file,"src/webmachine_resource.erl"},{line,125}]},{webmachine_decision_core,resource_call,1,[{file,"src/webmachine_decision_core.erl"},{line,48}]},{webmachine_decision_core,decision,1,[{file,"src/webmachine_decision_core.erl"},{line,416}]},{webmachine_decision_core,handle_request,2,[{file,"src/webmachine_decision_core.erl"},{line,33}]},{webmachine_mochiweb,loop,1,[{file,"src/webmachine_mochiweb.erl"},{line,97}]},{mochiweb_http,parse_headers,5,[{file,"src/mochiweb_http.erl"},{line,180}]},{proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,227}]}]
```

The docs in `www/resources.html` say that both `false` and `{halt, integer()}` returns are valid.

**Comment by slfritchie** [comment]
Pilot error, sorry.  The `{halt, integer()}` return works after I stop returning an invalid tuple from `delete_resource()`.  Criticism of `false` remains.

**Comment by beerriot** [comment]
I think this was fixed by #142. A few notes:

1. [The docs](https://github.com/webmachine/webmachine/wiki/Resource-Functions#delete_resource) were a bit obtuse about what the returns from `delete_resource` mean. Here, `false` means there was an error deleting the resource, so a 500 response is correct. I've made an update to the wiki that hopefully clarifies a bit.
2. Building `test_disco_resource` on near-current Webmachine (8413c977) does not close connections. I see every response come back as:

```
< HTTP/1.1 204 No Content
< Server: MochiWeb/2.20.0 WebMachine/1.11.1+build.627.ref8413c97 (greased slide to failure)
< Date: Tue, 31 Jan 2023 19:50:03 GMT
< Content-Type: text/html
< 
* Connection #0 to host localhost left intact
```
